pkg/controller/appdefinition/volume.go

```diff
@@ -206,6 +209,55 @@ func toPVCs(req router.Request, appInstance *v1.AppInstance) (result []kclient.O
 	return
 }
 
+func getPVForVolumeBinding(req router.Request, appInstance *v1.AppInstance, binding v1.VolumeBinding) (*corev1.PersistentVolume, error) {
+	// binding.Volume can either be the actual name of the PersistentVolume, or its public name in Acorn.
+	// Check for the actual name first.
+	pv := new(corev1.PersistentVolume)
+	if err := req.Client.Get(req.Ctx, kclient.ObjectKey{Name: binding.Volume}, pv); err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	} else if err == nil {
+		// make sure this PV is managed by acorn
+		if _, ok := pv.Labels[labels.AcornManaged]; !ok {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: "", Resource: "persistentvolumes"}, binding.Volume)
+		}
+
+		// make sure this PV has the Acorn app namespace label with the same value as the AppInstance's namespace
+		if appNamespace, ok := pv.Labels[labels.AcornAppNamespace]; !ok || appNamespace != appInstance.Namespace {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: "", Resource: "persistentvolumes"}, binding.Volume)
+		}
+		return pv, nil
+	}
+
+	// If we didn't find it by name, then look for it by public name.
+	// After the binding happens, the public name label will have been changed to the new AppInstance,
+	// so we also need to account for that in the label selector.
+	requirement, err := klabels.NewRequirement(labels.AcornPublicName, selection.In, []string{
+		binding.Volume,
+		publicname.ForChild(appInstance, binding.Target),
+	})
+	if err != nil {
+		return nil, err
+	}
+	selector := klabels.SelectorFromSet(map[string]string{
+		// also make sure we only list volumes that are part of the same Acorn project
+		labels.AcornAppNamespace: appInstance.Namespace,
+		labels.AcornManaged:      "true",
+	}).Add(*requirement)
+
+	pvList := new(corev1.PersistentVolumeList)
+	if err := req.Client.List(req.Ctx, pvList, &kclient.ListOptions{LabelSelector: selector}); err != nil {
+		return nil, err
+	}
+	if len(pvList.Items) != 1 {
+		if len(pvList.Items) == 0 {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: "", Resource: "persistentvolumes"}, binding.Volume)
+		}
+		return nil, fmt.Errorf("expected 1 PV for volume %s, found %d", binding.Volume, len(pvList.Items))
```

Comment:
What conditions will lead to this happening? Are we able to have more than one PV with the same public name?

Reply:
As far as I am aware, each PV public name should be unique within a project. I'm technically not even sure that this is possible, but I figured it would be best to check for that case anyway.

Reply:
Yeah, doesn't hurt. Just wanted to check 👍🏻